presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java

```diff
@@ -0,0 +1,234 @@
+    private final int maxFilterPositionsCount;
+    private final long maxFilterSizeInBytes;
+
+    private final List<Channel> channels;
+
+    // May be dropped if the predicate becomes too large.
+    private BlockBuilder[] blockBuilders;
+    private TypedSet[] valueSets;
+
+    private int filterPositionsCount;
+    private long filterSizeInBytes;
+
+    public DynamicFilterSourceOperator(OperatorContext context, Consumer<TupleDomain<String>> dynamicPredicateConsumer, List<Channel> channels, PlanNodeId planNodeId)
+    {
+        this.context = requireNonNull(context, "context is null");
+        this.maxFilterPositionsCount = getDynamicFilteringMaxPerDriverValueCount(context.getSession());
+        this.maxFilterSizeInBytes = getDynamicFilteringMaxPerDriverByteSize(context.getSession()).toBytes();
+
+        this.dynamicPredicateConsumer = requireNonNull(dynamicPredicateConsumer, "dynamicPredicateConsumer is null");
+        this.channels = requireNonNull(channels, "channels is null");
+
+        final int expectedSize = 8;
+        this.blockBuilders = new BlockBuilder[channels.size()];
+        this.valueSets = new TypedSet[channels.size()];
+        for (int channelIndex = 0; channelIndex < channels.size(); ++channelIndex) {
+            Type type = channels.get(channelIndex).type;
+            this.blockBuilders[channelIndex] = type.createBlockBuilder(null, expectedSize);
+            this.valueSets[channelIndex] = new TypedSet(type, Optional.empty(), blockBuilders[channelIndex], expectedSize,
+                                                        String.format("DynamicFilterSourceOperator_%s_%d", planNodeId, channelIndex));
+        }
+        this.filterPositionsCount = 0;  // non-null values count
+        this.filterSizeInBytes = 0;  // non-null values size in bytes
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return context;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return current == null && !finished;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        verify(!finished, "DynamicFilterSourceOperator: addInput() may not be called after finish()");
+        current = page;
+        if (valueSets == null) {
+            return;  // the predicate became too large.
+        }
+
+        // Collect only the columns which are relevant for the JOIN.
+        for (int channelIndex = 0; channelIndex < channels.size(); ++channelIndex) {
+            Block block = page.getBlock(channels.get(channelIndex).index);
+            TypedSet valueSet = valueSets[channelIndex];
+            filterSizeInBytes -= valueSet.getRetainedSizeInBytes();
+            for (int position = 0; position < block.getPositionCount(); ++position) {
+                if (!valueSet.contains(block, position)) {
+                    valueSet.add(block, position);
+                    filterPositionsCount += 1;
```

Comment:
types set won't count for duplicates, so you shouldn't either. Please add method `TypedSet#positionCount` and use it for counting

Reply:
IIUC, it can also be achieved by using the `TypedSet#size` count.